pkg/sidecar/sidecar_posix_test.go

```diff
@@ -0,0 +1,353 @@
+					t.Logf("sidecar reports it sent a signal: %+v", pidFileResult)
+					require.NoError(t, pidFileResult.err)
+				}
+			}
+			// Since a valid pid file was supplied, signalling must succeed
+			// on the first attempt
+			require.Equal(t, pid, pidFileResult.pid)
+			require.NoError(t, pidFileResult.err)
+
+			// Did we actually receive the signal?
+			select {
+			case <-ctx.Done():
+				// overall context has expired; this will fail the test.
+				require.NoError(t, ctx.Err())
+				return
+			case sig := <-sigListener:
+				require.Equal(t, tc.signal, sig)
+			}
+		})
+	}
+}
+
+// A long-running command should not be re-launched when the certs are rotated,
+// and must be signalled to reload the certs.
+//
+// This test could be amended to remove the signal handling part on Windows and
+// exercise a long running process by checking pids and chatting with it over
+// stdio but that's a task for another day.
+func TestSidecar_TestCmdRunsLongRunning(t *testing.T) {
+	const testsig = syscall.SIGUSR1
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	s := newSidecarTest(t)
+	s.NewConfig(t)
+	config := s.Config()
+	// Run a helper command that sleeps until signalled with SIGUSR1 then forwards
+	// the signal to the specified pid (this test process). It runs as an alternate
+	// mode of this test executable; see TestMain
+	config.Cmd = os.Args[0]
+	config.CmdArgs = fmt.Sprintf("%s %s %d", SignalHelperArg, SignalName(testsig), os.Getpid())
+	config.RenewSignal = SignalName(testsig)
+
+	log.Printf("will run: %+v", config)
+
+	s.NewSidecar(t)
+	sidecar := s.Sidecar()
+	defer s.Close(t)
+
+	// Listen for the signal we're going to send to the test helper, so we know when
+	// it gets forwarded to us
+	sigListener := make(chan os.Signal, 1)
+	signal.Notify(sigListener, testsig)
+	defer signal.Stop(sigListener)
+	t.Logf("listening for signal %s", SignalName(testsig))
+
+	// Fake the cert issue, which will start the process
+	svid := newTestX509SVID(t, s.rootCA)
+	s.MockUpdateX509Certificate(ctx, t, svid)
+
+	// It takes a moment for the test process to start up. We don't currently
+	// send a channel event for this, so we'll just busy-wait for the process
+	for {
+		sidecar.mu.Lock()
+		running := sidecar.processRunning
+		sidecar.mu.Unlock()
+		if running {
+			break
+		}
+		if ctx.Err() != nil {
+			require.NoError(t, ctx.Err())
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+	require.Equal(t, true, sidecar.processRunning)
+	firstPid := sidecar.process.Pid
+	fmt.Printf("sidecar process running: %d\n", firstPid)
+
+	// There should be no signal sent to the test helper on the first iteration
+	select {
+	case <-ctx.Done():
+		require.NoError(t, ctx.Err())
+	case <-sidecar.cmdExitChan:
+		require.Fail(t, "command should not have exited")
+	case <-time.After(100 * time.Millisecond):
+		// We should have received a signal by now if we were going to
+		// so all is well
```

Comment:
Should this be an error?

Reply:
No, it's expected to time out, since it is checking that no signal is received when none is expected. I can change the comment if needed.

Reply:
I've pushed a commit to improve this. It wasn't actually checking for receipt of the signal, which made the logic there confusing to read. Fixed.